### PR TITLE
perf: add column insert test

### PR DESCRIPTION
**Pull request by Gumix**
The test creates an empty space with 1000 nullable columns storing `uint64` values. Then it initializes a datasets that consists of 10 columns and 1 million rows (row count and both column counts are configurable), then it inserts the dataset into the space.

By default the test uses serial C API but one may switch to the Arrow API for batch insertion (the feature is exclusive to the Enterprise Edition).

It's also possible to specify the engine and `wal_mode` to use (default are `memtx`, `write`).

Needed for tarantool/tarantool-ee#712

### CODE REVIEW

File: perf/lua/CMakeLists.txt [hunk 1]
```diff
@@ -54,13 +54,19 @@ create_perf_lua_test(NAME gh-7089-vclock-copy)
 create_perf_lua_test(NAME uri_escape_unescape)
 
 include_directories(${MSGPUCK_INCLUDE_DIRS})
+include_directories(${EXTRA_BOX_INCLUDE_DIRS})
```
Comment (Gumix) [inline perf/lua/CMakeLists.txt:57]: Here I include [`third_party` from tarantool-ee](https://github.com/tarantool/tarantool-ee/blob/0a3e3d16754058a95dd56316650c68d8ed50fb73/CMakeLists.txt#L64). Not sure how to do it more correctly.

File: perf/lua/CMakeLists.txt [hunk 2]
```diff
@@ -54,13 +54,19 @@ create_perf_lua_test(NAME gh-7089-vclock-copy)
...
 target_link_libraries(column_scan_module msgpuck)
 create_perf_lua_test(NAME column_scan
                      DEPENDS column_scan_module
 )
 
+build_module(column_insert_module column_insert_module.c)
+target_link_libraries(column_insert_module msgpuck)
+create_perf_lua_test(NAME column_insert
+                     DEPENDS column_insert_module
+)
```
Comment (mkostoevr) [inline perf/lua/CMakeLists.txt:69]: Just a sidenote: a newline would be nice here.

File: perf/lua/column_insert.lua [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    {'wal_mode', 'string'},
+    {'column_count_total', 'number'},
+    {'column_count_batch', 'number'},
+    {'row_count_total', 'number'},
+    {'row_count_batch', 'number'},
+    {'use_arrow_api', 'boolean'},
+}, USAGE)
+
+local DEFAULT_ENGINE = 'memtx'
+local DEFAULT_WAL_MODE = 'none'
```
Comment (mkostoevr) [inline perf/lua/column_insert.lua:40]: Seems reasonable to make it "write" by default (the same way it's "memtx" instead of "memcs" by default).

Just toughts, feel free to ignore this.

File: perf/lua/column_insert_module.c [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+	for (int64_t i = 0; i < dataset.row_count; i++) {
+		char *data_end = tuple_data;
+		data_end = mp_encode_array(data_end, total_column_count);
+		int j;
+		for (j = 0; j < dataset.column_count; j++) {
+			uint64_t val = dataset.columns[j].data[i];
+			data_end = mp_encode_uint(data_end, val);
+		}
+		for (; j < total_column_count; j++)
+			data_end = mp_encode_nil(data_end);
```
Comment (mkostoevr) [inline perf/lua/column_insert_module.c:45]: Just a question, is it known how does the creation of tuples prior to each insertion affect the benchmarking results?

I see creating tuples beforehand could require a whole bunch of memory.

**Review by mkostoevr — APPROVED** [review]
LGTM, just a few notes.

File: extra/exports [hunk 5]
```diff
@@ -265,6 +265,7 @@ luaL_pushcdata
 luaL_pushint64
 luaL_pushuint64
 luaL_setcdatagc
+luaL_tocpointer
```
Comment (locker) [inline extra/exports:268]: AFAICS you use it to implement `luaL_isnull` in tests for https://github.com/tarantool/tarantool-ee/pull/830. I think it's better to make `luaL_isnull` and `luaL_pushnull` public and do it in a separate commit.

File: perf/lua/CMakeLists.txt (hunk 1 above)
Comment (locker) [reply perf/lua/CMakeLists.txt:57]: I don't like including extra box directories. Let's define `ARROW_INCLUDE_DIRS` in Tarantool EE.

File: perf/lua/column_insert.lua [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+assert(params.row_count_total % params.row_count_batch == 0)
+
+local bench = benchmark.new(params)
+
+local BUILDDIR = fio.abspath(fio.pathjoin(os.getenv('BUILDDIR') or '.'))
+local MODULEPATH = fio.pathjoin(BUILDDIR, 'perf', 'lua',
+                                '?.' .. tarantool.build.mod_format)
+package.cpath = MODULEPATH .. ';' .. package.cpath
+
+local test_module = require('column_insert_module')
```
Comment (ligurio) [inline perf/lua/column_insert.lua:68]: Please call this `require` under `pcall` and print user-friendly error if module is not available.

File: perf/lua/column_insert_module.c (hunk 4 above)
Comment (Gumix) [reply perf/lua/column_insert_module.c:45]: I think it is more fair to include the tuple creation time into the total measured time. Anyway, it is negligible. E.g. for 100 columns:

Screenshot 2024-07-08 at 15 11 08

File: perf/lua/CMakeLists.txt (hunk 1 above)
Comment (Gumix) [reply perf/lua/CMakeLists.txt]: Removed this line and included `third_party/arrow/abi.h` into CE.

File: perf/lua/column_insert.lua (hunk 3 above)
Comment (Gumix) [reply perf/lua/column_insert.lua]: Done, with the following changed:
```diff
 box.cfg({
     log = 'tarantool.log',
+    work_dir = test_dir,
     wal_mode = params.wal_mode,
     memtx_memory = 4 * 1024 * 1024 * 1024,
     checkpoint_count = 1,
@@ -144,4 +166,5 @@ end
 bench:dump_results()

 test_module.fini()
+rmtree(test_dir)
 os.exit(0)
```

```diff
@@ -28,11 +28,14 @@ insert_serial_lua_func(struct lua_State *L)
 {
        uint32_t space_id = luaL_checkinteger(L, 1);
        luaL_checktype(L, 2, LUA_TTABLE);
+       lua_getfield(L, 2, "row_count_batch");
+       int batch_row_count = luaL_checkinteger(L, -1);
        lua_getfield(L, 2, "column_count_total");
        int total_column_count = luaL_checkinteger(L, -1);
-       lua_pop(L, 1);
+       lua_pop(L, 2);
        static char tuple_data[1000 * 1000];

+       VERIFY(box_txn_begin() == 0);
        for (int64_t i = 0; i < dataset.row_count; i++) {
                char *data_end = tuple_data;
                data_end = mp_encode_array(data_end, total_column_count);
@@ -48,7 +51,12 @@ insert_serial_lua_func(struct lua_State *L)
                        abort();
                if (box_insert(space_id, tuple_data, data_end, NULL) != 0)
                        return luaT_error(L);
+               if (i % batch_row_count == 0) {
+                       VERIFY(box_txn_commit() == 0);
+                       VERIFY(box_txn_begin() == 0);
+               }
        }
+       VERIFY(box_txn_commit() == 0);
        return 0;
 }
```

File: extra/exports (hunk 5 above)
Comment (Gumix) [reply extra/exports]: Done.

File: perf/lua/column_insert.lua (hunk 6 above)
Comment (Gumix) [reply perf/lua/column_insert.lua]: Done:
```diff
-local test_module = require('column_insert_module')
-local test_funcs = {}
+local test_module_name = 'column_insert_module'
+local has_test_module, test_module = pcall(require, test_module_name)
+if not has_test_module then
+    local errmsg = ('Lua module "%s" is not found.\n'):format(test_module_name)
+    io.stderr:write(errmsg)
+    os.exit(1)
+end

+local test_funcs = {}
```

File: third_party/arrow/abi.h
Comment (locker) [inline third_party/arrow/abi.h:1]: Please don't forget to drop the copy from the EE repository.

**Review by locker — APPROVED** [review]

File: changelogs/unreleased/expose-luaL_pushnull-and-luaL_isnull [hunk 7]
```diff
@@ -0,0 +1,3 @@
+## feature/lua
+
+* Exposed `luaL_pushnull()` and `luaL_isnull()` functions via C module API.
```
Comment (andreyaksenov) [inline changelogs/unreleased/expose-luaL_pushnull-and-luaL_isnull:3]:
```suggestion
* Exposed the `luaL_pushnull()` and `luaL_isnull()` functions via the C module API.
```

**Review by andreyaksenov — APPROVED** [review]

File: changelogs/unreleased/expose-luaL_pushnull-and-luaL_isnull (hunk 7 above)
Comment (Gumix) [reply changelogs/unreleased/expose-luaL_pushnull-and-luaL_isnull]: Done.

**Comment by locker** [comment]
@Gumix CI fails, please fix. In future, please run `full-ci` by yourself after soliciting all approvals before assigning a PR to me for merge.

**Comment by Gumix** [comment]
> @Gumix CI fails, please fix.

1. `integration / mysql` fixed by:

- https://github.com/tarantool/mysql/pull/76

2. `default_gcc_centos_7 / default_gcc_centos_7` fixed by [adding missed compilation flags](https://github.com/tarantool/tarantool/compare/781ed965c306cf956064d9d853bf2ed676b47bd4..b2f78c7011da29364229ed2f81326ac1455157fa), similar to:
https://github.com/tarantool/tarantool/blob/3e1830447c5b1fb018966586ffbd3db080460f5b/test/CMakeLists.txt#L1-L2
